mesonbuild/interpreter/compiler.py

```diff
@@ -743,3 +747,59 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
     @noKwargs
     def get_argument_syntax_method(self, args: T.List['TYPE_var'], kwargs: 'TYPE_kwargs') -> str:
         return self.compiler.get_argument_syntax()
+
+    @typed_kwargs('compiler.symbol_visibility_header',
+        KwargInfo('api', (str, NoneType)),
+        KwargInfo('compilation', (str, NoneType)),
+        KwargInfo('static_compilation', (str, NoneType)),
+        KwargInfo('static_only', bool, default=False),
+        )
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '0.63.0')
+    def symbol_visibility_header_method(self, args: T.List[str], kwargs: 'TYPE_kwargs') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp'}:
+            raise InterpreterException('compiler.symbol_visibility_header is supported only for C and C++ compilers')
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        if default_prefix[0].isdigit():
+            default_prefix = f'_{default_prefix}'
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = ''
+
+        if static_only:
+            content += textwrap.dedent(f'''\
+                #ifndef {static_compilation}
+                #  define {static_compilation}
+                #endif /* {static_compilation} */
+                ''')
+
+        content += textwrap.dedent(f'''\
+            #ifndef {api}
+            #  if defined _WIN32 || defined __CYGWIN__
+            #    ifdef {static_compilation}
+            #      define {api} extern
+            #    else /* !{static_compilation} */
+            #      ifdef {compilation}
+            #        define {api} extern __declspec(dllexport)
+            #      else /* !{compilation} */
+            #        define {api} extern __declspec(dllimport)
+            #      endif /* !{compilation} */
+            #    endif /* !{static_compilation} */
+            #  elif __GNUC__ >= 4
```

Comment:
Feels like a trap to me when all compilers supported by Meson can tell you if they support this or not, but if others don't agree then it can just be fixed in a future bug report.

Reply:
That's what says GNU's own documentation...